dedup/src/main.rs

```diff
@@ -0,0 +1,119 @@
+use std::fs::File;
+use std::path::{Path,PathBuf};
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+use std::io::{self, Read};
+
+extern crate md5;
+extern crate walkdir;
+
+const FOLLOW_SYMLINKS: bool = false;
+const ROOT_DIRECTORY: &'static str = "/";
+//const ROOT_DIRECTORY: &'static str = "/home/owen";
+//const REPORT_ERRORS: bool = false;
+const BUFFER_SIZE: usize = 8096;   // 20MB  // TODO: tweak
+
+type FileIter = Box<Iterator<Item=walkdir::DirEntry>>;
+
+/*
+enum SelectionStrat {
+    PathLen,
+    CreationTime,
+    UpdateTime,
+    //Drive(String),
+    Prefix(String),
+    NotPrefix(String),
+}
+type SelectionStrats = Vec<SelectionStrat>;
+
+enum ResolutionStrat {
+    Nothing,
+    Delete,
+    Symlink,
+}
+*/
+
+fn get_info(path: &Path) -> io::Result<(u64,[u8;16])> {
+    let mut f = File::open(path)?;
+    let mut buffer: Box<[u8]> = vec![0u8; BUFFER_SIZE].into_boxed_slice();
+    let mut total = 0usize;
+    let mut context = md5::Context::new();
+    loop {
+        let size = f.read(&mut buffer)?;
+        if size == 0 {
+            let digest = context.compute();
+            return Ok((total as u64, *digest));
+        }
+        total += size;
+        context.consume(&buffer[..size]);
+        for b in buffer.iter_mut() { *b = 0; }  // maybe use mem::zeroed?
```

Comment:
https://doc.rust-lang.org/book/first-edition/iterators.html

If `x` is a collection of type `T` then `i` will be of type `T`. If `T` is not a primitive/`Copy` type (e.g. `Vec<String>`), each `i` and will take ownership and destroy the value when it goes out of scope. So `for i in x {} for i in x {}` is a compile time error because `x` has been consumed.
Instead you can call `for i in &x{}` in which case `i` will be of type `&T` (an immutable reference) and you can iterate many times. Here I'm taking a mutable reference.